conda_declarative/state.py

```diff
@@ -0,0 +1,122 @@
+"""Code for managing the state of env.yml."""
+
+import pathlib
+from collections.abc import Iterable
+from dataclasses import asdict
+
+from conda.base.context import env_name
+from conda.common.serialize import yaml_safe_dump, yaml_safe_load
+from conda.history import History
+from conda.models.enums import Arch, Platform
+from conda.models.environment import Environment, EnvironmentConfig
+from conda.models.match_spec import MatchSpec
+
+
+def get_platform() -> str:
+    """Get the current platform.
+
+    Right now there's no way to get the current platform, so we roll our own function
+    to do this here. Yes, you can use `conda.models.enums.Platform.from_sys`, but the
+    values of this enum are not valid to pass to `conda.models.environment.Environment`,
+    which must be one of `conda.base.constants.PLATFORMS`.
+
+    Returns
+    -------
+    str
+        Currently running platform; one of conda.base.constants.PLATFORMS
+    """
+    arch_enum = Arch.from_sys()
+    match arch_enum:
+        case Arch.x86:
+            arch = "32"
+        case Arch.x86_64:
+            arch = "64"
+        case _:
+            arch = arch_enum.value
+
+    platform_enum = Platform.from_sys()
+    match platform_enum:
+        case Platform.osx:
+            platform = "osx"
+        case Platform.win:
+            platform = "win"
+        case _:
+            platform = platform_enum.value
+
+    return f"{platform}-{arch}"
+
+
+def update_state(
+    prefix: str | None,
+    remove_specs: Iterable[MatchSpec] | None = None,
+    update_specs: Iterable[MatchSpec] | None = None,
+) -> None:
+    """Update `conda-meta/env.yml` with the current packages in the environment.
+
+    The requested packages in the environment are read from the existing env file.
+    If no env file exists, the requested packages are read from the history file.
+
+    Parameters
+    ----------
+    prefix : str | None
+        Prefix for which is being modified by a user command
+    remove_specs : Iterable[MatchSpec] | None
+        Packages the user has requested to remove
+    update_specs : Iterable[MatchSpec] | None
+        Packages the user has either requested to add or update
+    """
+    if get_env_file(prefix).exists():
+        current_env = get_environment(prefix)
+
+        packages = {}
+        for pkg in current_env.requested_packages:
+            if pkg not in remove_specs:
+                packages[pkg.name] = pkg
+
+    else:
+        packages = {}
+        for pkg in History(prefix=prefix).get_requested_specs_map().values():
+            if pkg not in remove_specs:
+                packages[pkg.name] = pkg
+
+    packages.update({pkg.name: pkg for pkg in update_specs})
+    env_dict = asdict(
+        Environment(
+            prefix=str(prefix),
+            platform=get_platform(),
+            config=EnvironmentConfig(),
```

Comment:
All the `EnvironmentConfig` fields should be able to be populated from the context object. There is a 1-to-1 relationship between the fields in this config and the context.

Reply:
Cool, that's helpful. Kinda makes sense too, as the `context` is sort of the environment configuration for whatever environment the user is currently looking at.